src/metabase/api/collection.clj

```diff
@@ -541,16 +542,65 @@
   [_ collection options]
   (collection-query collection options))
 
+(defn- annotate-collections
+  [parent-coll colls]
+  (let [visible-collection-ids (collection/permissions-set->visible-collection-ids
+                                @api/*current-user-permissions-set*)
+
+        child-type->coll-id-set
+        (reduce (fn [acc {collection-id :collection_id, card-type :type, :as _card}]
+                  (update acc (if (= (keyword card-type) :model) :dataset :card) conj collection-id))
+                {:dataset #{}
+                 :card    #{}}
+                (mdb.query/reducible-query {:select-distinct [:collection_id :type]
+                                            :from            [:report_card]
+                                            :where           [:= :archived false]}))
```

Comment:
I'm a little concerned about the performance here. This is fetching _all_ collections on the instance that contain cards, dashboards, or collections, just to annotate `:here` and `:below` on a subset of collections passed into this function?

That seems like it's doing a lot more work than necessary. Could there be issues if the instance has a huge number of collections? Can we scope these queries to just the collections that are relevant?

(If the answer is no, and I'm misunderstanding something, maybe a docstring/comment could help to explain why for other readers)

Reply:
Hmm, I think we could actually - we could get the *collections* that are descendants of the parent first. Then add `[:in :collection_id (map :id descendants)]` to the conditions here (and for the similar query selecting `collection_id` from dashboards below).

Worth noting that we do something very similar for the `/tree` endpoint, so I think so far this hasn't been a problem for performance. But it made me uncomfortable too, I just hadn't thought of a good solution until now.